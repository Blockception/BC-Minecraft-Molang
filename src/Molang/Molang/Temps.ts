import { RegularExpression } from "../RegExp";

/**The namespace that governs molang temp variables*/
export namespace Temps {
  /**The pattern used to find the defintions of temp variables*/
  export const getDefinedPatt: RegExp = /(?:^|;[ \t]*|"[ \t]*)\b(?:t|temp)\.([a-z0-9_]*)[ \t]*\=/gim;

  /**The pattern used to find the temp variables*/
  export const getUsedPatt: RegExp = /\b(?:t|temp)\.([a-z0-9_]+)\b(?![ \t]+=)/gim;

  /**Spits through all the provided strings searching for all instances of defining temp variables.
   * @param data The string(s)/container to look through
   * @param receiver The receiving array*/
  export function getDefined(data: any, receiver: string[]): void {
    RegularExpression.harvest(data, getDefinedPatt, receiver);
  }

  /**Spits through all the provided strings searching for all instances of using temp variables.
   * @param data The string(s)/container to look through
   * @param receiver The receiving array*/
  export function getUsing(data: any, receiver: string[]): void {
    RegularExpression.harvest(data, getUsedPatt, receiver);
  }
}
