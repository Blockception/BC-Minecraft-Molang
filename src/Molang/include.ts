/*	Auto generated	*/

export * as Molang from "./Molang/include";
export * from "./MolangSet";
export * from "./RegExp";
