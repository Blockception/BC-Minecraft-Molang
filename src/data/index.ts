/*	Auto generated	*/
export * from "./animations";
export * from "./animations-controllers";
export * from "./attachables";
export * from "./blocks";
export * from "./data";
export * from "./entities";
export * from "./features-rules";
export * from "./general";
export * from "./items";
export * from "./molang-data";
export * from "./molang-function";
export * from "./particles";
export * from "./render-controllers";

