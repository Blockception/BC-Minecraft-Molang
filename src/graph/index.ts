export * as Mermaid from './mermaid'