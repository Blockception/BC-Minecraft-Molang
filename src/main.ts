export * from "./data";
export * as Molang from "./molang";
export * from "./types/defined";
