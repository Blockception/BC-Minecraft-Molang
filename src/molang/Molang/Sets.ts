import { RegularExpression } from "../RegExp";
import { MolangFullSet, MolangSet } from "../MolangSet";

/**The namespace that governs molang sets*/
export namespace Sets {
  /**The pattern used to find the used to find most items*/
  export const getUsedPatt: RegExp = /\b(c|context|t|temp|q|query|m|math|v|variable)\.([a-z0-9_]+)\b(?![ \t]+=)/gim;
  /**The pattern used to find the used to find all items*/
  export const getFullUsedPatt: RegExp =
    /\b((t|temp|v|variable)\.([a-z0-9_]+)\b(?![ \t]+=)|(c|context|q|query|m|math|texture|geometry|material|array)\.([a-z0-9_][a-z0-9_\.]+)\b)/gim;
  /**The pattern used to find the used to find all defined items*/
  export const getDefinedPatt: RegExp = /(?:^|;[ \t]*|"[ \t]*)\b(v|t|variable|temp)\.([a-z0-9_]*)[ \t]*\=/gim;

  /**Spits through all the provided strings searching for all instances of using textures
   * @param data The string(s)/container to look through
   * @param receiver The receiving array*/
  export function getUsingFull(data: any, set: MolangFullSet): void {
    const receiver: string[] = [];
    RegularExpression.harvest(data, getFullUsedPatt, receiver);

    console.log(JSON.stringify(receiver));

    for (let I = 0; I < receiver.length; I++) {
      const type = receiver[I++];
      const value = receiver[I];

      if (type && value) {
        switch (type.toLowerCase()) {
          case "c":
          case "context":
            set.contexts.using.push(value);
            break;

          case "v":
          case "variable":
            set.variables.using.push(value);
            break;

          case "t":
          case "temp":
            set.temps.using.push(value);
            break;

          case "q":
          case "query":
            set.queries.using.push(value);
            break;

          case "math":
            //TODO
            break;

          case "texture":
            set.textures.using.push(value);
            break;

          case "geometry":
            set.geometries.using.push(value);
            break;

          case "material":
            set.materials.using.push(value);
            break;
        }
      }
    }
  }

  /**Spits through all the provided strings searching for all instances of using textures
   * @param data The string(s)/container to look through
   * @param receiver The receiving array*/
  export function getUsing(data: any, set: MolangSet): void {
    const receiver: string[] = [];
    RegularExpression.harvest(data, getUsedPatt, receiver);

    for (let I = 0; I < receiver.length; I++) {
      const type = receiver[I++];
      const value = receiver[I];

      if (type && value) {
        switch (type.toLowerCase()) {
          case "c":
          case "context":
            set.contexts.using.push(value);
            break;

          case "v":
          case "variable":
            set.variables.using.push(value);
            break;

          case "t":
          case "temp":
            set.temps.using.push(value);
            break;

          case "q":
          case "query":
            set.queries.using.push(value);
            break;

          case "math":
            //TODO
            break;
        }
      }
    }
  }

  /**Spits through all the provided strings searching for all instances of using textures
   * @param data The string(s)/container to look through
   * @param receiver The receiving array*/
  export function getDefined(data: any, set: MolangSet): void {
    const receiver: string[] = [];
    RegularExpression.harvest(data, getDefinedPatt, receiver);

    for (let I = 0; I < receiver.length; I++) {
      const type = receiver[I++];
      const value = receiver[I];

      if (type && value) {
        switch (type.toLowerCase()) {
          case "v":
          case "variable":
            set.variables.defined.push(value);
            break;

          case "t":
          case "temp":
            set.temps.defined.push(value);
            break;
        }
      }
    }
  }
}
