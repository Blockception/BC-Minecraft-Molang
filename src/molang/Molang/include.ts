/*	Auto generated	*/

export * from "./Context";
export * from "./Geometries";
export * from "./Materials";
export * from "./Queries";
export * from "./Temps";
export * from "./Textures";
export * from "./Variables";
