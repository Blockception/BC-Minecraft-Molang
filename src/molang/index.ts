/*	Auto generated	*/
export * as Types from "./types/index";
export * from "./functions";
export * from "./molang-set";
export * from "./regexp";
export * from "./syntax";
export * from "./traverse";