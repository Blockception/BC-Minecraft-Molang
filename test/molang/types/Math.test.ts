import { Types } from "../../../src/Molang";
import { VanillaPlayer } from "../../vanilla-player";

describe("Molang", () => {
  describe("math", () => {
    it("using", () => {
      const receiver: string[] = [];
      Types.Math.getUsing(VanillaPlayer.Data, receiver);
      expect(receiver).toMatchSnapshot();
    });

    it("using obj", () => {
      const receiver: string[] = [];
      Types.Math.getUsing(VanillaPlayer.DataOBject, receiver);
      expect(receiver).toMatchSnapshot();
    });
  });
});
