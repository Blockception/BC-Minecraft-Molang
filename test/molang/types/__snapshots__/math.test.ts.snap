// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Molang math using 1`] = `
[
  "sin",
  "sqrt",
  "cos",
  "clamp",
  "pow",
]
`;

exports[`Molang math using obj 1`] = `
[
  "sin",
  "sqrt",
  "cos",
  "clamp",
  "pow",
]
`;
